integration/version_test.go

```diff
@@ -0,0 +1,258 @@
+				return nil, trace.Wrap(err)
+			}
+			signal <- struct{}{}
+
+			return proc, nil
+		})
+	}()
+
+	timeout := time.After(time.Second * 30)
+	select {
+	case err := <-authRunErrCh:
+		require.NoError(t, err)
+	case <-timeout:
+		t.Fatal("timed out waiting for fetching identity")
+	case <-signal:
+		localVersion, err := processStorage.GetTeleportVersion(ctx)
+		require.NoError(t, err)
+
+		assert.Equal(t, teleport.Version, localVersion)
+	}
+}
+
+// TestVersionCheckUpgradeOneMajorVersion tests normal flow of major version upgrade,
+// should verify that after launch local version is upgraded as well.
+func TestVersionCheckUpgradeOneMajorVersion(t *testing.T) {
+	lib.SetInsecureDevMode(true)
+	defer lib.SetInsecureDevMode(false)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// Create temporary directory for the auth service.
+	authDir := t.TempDir()
+
+	processStorage, err := storage.NewProcessStorage(ctx, filepath.Join(authDir, teleport.ComponentProcess))
+	require.NoError(t, err)
+
+	version := semver.New(teleport.Version)
+	err = processStorage.WriteTeleportVersion(fmt.Sprintf("%d.0.0", version.Major-1))
+	require.NoError(t, err)
+
+	authAddr, err := getFreeListenAddr()
+	require.NoError(t, err)
+
+	authCfg := servicecfg.MakeDefaultConfig()
+	authCfg.Version = defaults.TeleportConfigVersionV3
+	authCfg.DataDir = authDir
+	require.NoError(t, authCfg.SetAuthServerAddresses([]utils.NetAddr{
+		{
+			AddrNetwork: "tcp",
+			Addr:        authAddr,
+		},
+	}))
+	authCfg.Auth.Enabled = true
+	authCfg.Auth.StorageConfig.Params = backend.Params{defaults.BackendPath: filepath.Join(authDir, defaults.BackendDir)}
+	authCfg.Auth.ClusterName, err = services.NewClusterNameWithRandomID(types.ClusterNameSpecV2{
+		ClusterName: "auth-server",
+	})
+	require.NoError(t, err)
+	authCfg.Auth.ListenAddr.Addr = authAddr
+	authCfg.Auth.NetworkingConfig.SetProxyListenerMode(types.ProxyListenerMode_Multiplex)
+
+	authCfg.Proxy.Enabled = false
+	authCfg.SSH.Enabled = false
+
+	authRunErrCh := make(chan error, 1)
+	signal := make(chan struct{})
+	go func() {
+		authRunErrCh <- service.Run(ctx, *authCfg, func(cfg *servicecfg.Config) (service.Process, error) {
+			proc, err := service.NewTeleport(cfg)
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+
+			if _, err = proc.GetIdentity(types.RoleAdmin); err != nil {
+				proc.Close()
+				return nil, trace.Wrap(err)
+			}
+			signal <- struct{}{}
+
+			return proc, nil
+		})
+	}()
+
+	timeout := time.After(time.Second * 30)
+	select {
+	case err := <-authRunErrCh:
+		require.NoError(t, err)
+	case <-timeout:
+		t.Fatal("timed out waiting for fetching identity")
+	case <-signal:
+		localVersion, err := processStorage.GetTeleportVersion(ctx)
+		require.NoError(t, err)
+
+		assert.Equal(t, teleport.Version, localVersion)
+	}
+}
+
+// TestVersionCheckUpgradeFailed tests version upgrade restriction with incompatible major upgrade.
+func TestVersionCheckUpgradeFailed(t *testing.T) {
```

Comment:
Looks like these 3 tests share a lot of common code. Can we turn them into table tests?